main.cpp

```diff
@@ -96,224 +98,448 @@ inf
+//------------
+
+DoubleType& DoubleType::add(double rhs)
 {
-    return lhs * rhs;
+    *value += rhs;
+    return *this;
 }
 
-double DoubleType::divide(double lhs, double rhs)
+DoubleType& DoubleType::subtract(double rhs)
+{
+    *value -= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(double rhs)
+{
+    *value *= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(double rhs)
 {
     if(rhs == 0.0)
     {
-        std::cout << "\nwarning, floating point division by zero returns 'inf' !" << std::endl;
+        std::cout << "warning: floating point division by zero!" << std::endl;
     }
-    return lhs / rhs;
+    if(value != nullptr) *value /= rhs;
+    return *this;
 }
 
-struct IntType
+DoubleType& DoubleType::add(const FloatType& ft)
 {
-    int add(int lhs, int rhs);
-    int subtract(int lhs, int rhs);
-    int multiply(int lhs, int rhs);
-    int divide(int lhs, int rhs);
-};
+    return add(*ft.value);
+}
 
-int IntType::add(int lhs, int rhs)
+DoubleType& DoubleType::subtract(const FloatType& ft)
 {
-    return lhs + rhs;
+    return subtract(*ft.value);
 }
 
-int IntType::subtract(int lhs, int rhs)
+DoubleType& DoubleType::multiply(const FloatType& ft)
 {
-    return lhs - rhs;
+    return multiply(*ft.value);
 }
 
-int IntType::multiply(int lhs, int rhs)
+DoubleType& DoubleType::divide(const FloatType& ft)
 {
-    return lhs * rhs;
+    return divide(*ft.value);
 }
 
-int IntType::divide(int lhs, int rhs)
+DoubleType& DoubleType::add(const DoubleType& dt)
+{
+    return add(*dt.value);
+}
+
+DoubleType& DoubleType::subtract(const DoubleType& dt)
+{
+    return subtract(*dt.value);
+}
+
+DoubleType& DoubleType::multiply(const DoubleType& dt)
+{
+    return multiply(*dt.value);
+}
+
+DoubleType& DoubleType::divide(const DoubleType& dt)
+{
+    return divide(*dt.value);
+}
+
+DoubleType& DoubleType::add(const IntType& it)
+{
+    return add(*it.value);
+}
+
+DoubleType& DoubleType::subtract(const IntType& it)
+{
+    return subtract(*it.value);
+}
+
+DoubleType& DoubleType::multiply(const IntType& it)
+{
+    return multiply(*it.value);
+}
+
+DoubleType& DoubleType::divide(const IntType& it)
+{
+    return divide(*it.value);
+}
+
+//-----------
+
+IntType& IntType::add(int rhs)
+{
+    *value += rhs;
+    return *this;
+}
+
+IntType& IntType::subtract(int rhs)
+{
+    *value -= rhs;
+    return *this;
+}
+
+IntType& IntType::multiply(int rhs)
+{
+    *value *= rhs;
+    return *this;
+}
+
+IntType& IntType::divide(int rhs)
 {
     if(rhs == 0)
     {
-        std::cout << "error, integer division by zero will crash the program!" << std::endl;
-        std::cout << "returning lhs" << std::endl;
-        return lhs;
-    }
-    else
-    {
-        return lhs / rhs;
+        std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl;
+        return *this;
     }
+    if(value != nullptr) *value /= rhs;
+    return *this;
```

Comment:
Be consistent with your coding style for `if` expressions.
You've got curly braces and the `if` body on its own line in the first `if` expression
but the 2nd one doesn't have curly braces, nor is the `if` body on its own line.
Pick one style or the other
- [ ] revise